README.md
# ChatRPC

[![NPM version](https://img.shields.io/npm/v/chatrpc.svg)](https://www.npmjs.com/package/chatrpc)
[![Codecov](https://img.shields.io/codecov/c/github/floriscornel/ChatRPC.svg)](https://codecov.io/gh/floriscornel/ChatRPC)
[![GitHub Action](https://img.shields.io/github/actions/workflow/status/floriscornel/ChatRPC/check-main.yaml)](https://github.com/floriscornel/ChatRPC/actions/workflows/check-main.yaml)
[![License](https://img.shields.io/github/license/floriscornel/ChatRPC.svg)](https://github.com/floriscornel/ChatRPC/blob/main/LICENSE)
[![GitHub stars](https://img.shields.io/github/stars/floriscornel/ChatRPC.svg?style=social&logo=github&label=Stars)](https://github.com/floriscornel/ChatRPC)

ChatRPC is a framework that allows large language models to interact with external services.

Users of the library define methods and services:
```typescript
const isEven = new Method({
  handler: async (input: number) => input % 2 === 0,
  input: { type: 'number' },
  output: { type: 'boolean' },
});

const calculatorService = new Service({ name: 'calculator' })
  .registerMethod('isEven',isEven);
```

The LLM's will be prompted to output only valid JSON that can either contain a `message` to the user or a `method` call, for example:
```
User: {"message":"Hello, I would like to get information about currency conversion."}
Assistant: {"message":"I have have access to a realtime currency conversion service. What currency would you like to convert from?"}
User: {"message":"I would like to convert 500 USD to EUR."}
Assistant: {"service":"currency","method":"convert","input":{"amount":500,"from":"USD","to":"EUR"}}
System: {"output":415.53}
Assistant: {"message":"500 USD is equivalent to 415.53 EUR."}
```

## Example Application
A working example that integrates TMDB into ChatGPT can be found here: [examples/tmdb-openai](https://github.com/floriscornel/ChatRPC/tree/main/examples/tmdb-openai).

<img src="https://floriscornel.nl/ChatRPC-tmdb-demo.gif" width="655" height="422">

<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE.txt` for more information.


## Acknowledgments

* [typescript-starter](https://github.com/bitjson/typescript-starter) has helped me get started with this project.
  

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.1.3](https://github.com/floriscornel/ChatRPC/compare/v0.1.2...v0.1.3) (2023-06-14)

### Changed
- Updated dependencies

## [0.1.2](https://github.com/floriscornel/ChatRPC/compare/v0.1.1...v0.1.2) (2023-04-02)

### Changed
- Improved documentation and code coverage
- Reduce package size by removing unnecessary files

## [0.1.1](https://github.com/floriscornel/ChatRPC/compare/v0.1.0...v0.1.1) (2023-04-02)

### Changed
- Fixed validation function bug
 
## [0.1.0](https://github.com/floriscornel/ChatRPC/commits/v0.1.0) (2023-04-02)

### Added
- Initial release of this library
- Example usage project in [examples/tmdb-openai](examples/tmdb-openai)
- Empty project based on typescript-standard-template 


src/lib/openai.ts
import { messageWrapperSchema, methodRequestWrapperSchema } from './interface';

/**
 * This is a builtin prompt that can be used when creating a new {@link Chat}.
 * This prompt was tested with the GPT-3 model of OpenAI.
 *
 * @example
 * ```
 * Pretend to be a machine that can interact with external services. All messages must be in JSON format.
 *
 * Rules:
 * 1. All messages to the User must be in the following format:
 * {"type":"object","properties":{"message":{"type":"string"}},"required":["message"]}
 *
 * 3. You can request the execution of a method of a service by sending the following format:
 * {"type":"object","properties":{"service":{"type":"string"},"method":{"type":"string"},"input":{"type":"any"}},"required":  * ["service","method","input"]}
 *
 * Note: The user cannot see the output of the method execution. The output is only visible to the system.
 * You must inform the user of the output of the method execution.
 * To get the available services and methods, send the following message:
 * Assistant: {"service":"root","method":"getServices"}
 *
 * A typical conversation might look like this:
 * User: {"message":"Hello, I would like to get information about currency conversion."}
 * Assistant: {"message":"I have have access to a realtime currency conversion service. What currency would you like to  convert from?"}
 * User: {"message":"I would like to convert 500 USD to EUR."}
 * Assistant: {"service":"currency","method":"convert","input":{"amount":500,"from":"USD","to":"EUR"}}
 * System: {"output":415.53}
 * Assistant: {"message":"500 USD is equivalent to 415.53 EUR."}
 *
 * Today is 4/2/2023.
 * ```
 */
export const getChatGpt3Prompt = (): string => {
  return `Pretend to be a machine that can interact with external services. All messages must be in JSON format.

Rules:
1. All messages to the User must be in the following format:
${JSON.stringify(messageWrapperSchema)}

3. You can request the execution of a method of a service by sending the following format:
${JSON.stringify(methodRequestWrapperSchema)}

Note: The user cannot see the output of the method execution. The output is only visible to the system.
You must inform the user of the output of the method execution.
To get the available services and methods, send the following message:
Assistant: ${JSON.stringify({ service: 'root', method: 'getServices' })}

A typical conversation might look like this:
User: ${JSON.stringify({
    message:
      'Hello, I would like to get information about currency conversion.',
  })}
Assistant: ${JSON.stringify({
    message:
      'I have have access to a realtime currency conversion service. What currency would you like to convert from?',
  })}
User: ${JSON.stringify({ message: 'I would like to convert 500 USD to EUR.' })}
Assistant: ${JSON.stringify({
    service: 'currency',
    method: 'convert',
    input: { amount: 500, from: 'USD', to: 'EUR' },
  })}
System: ${JSON.stringify({ output: 415.53 })}
Assistant: ${JSON.stringify({
    message: '500 USD is equivalent to 415.53 EUR.',
  })}

Today is ${new Date().toLocaleDateString()}.`;
};
